File: go/problems/pg42861/solution.go
// Package pg42861 - [PG] 42861 - 섬 연결하기
// https://programmers.co.kr/learn/courses/30/lessons/42861
// 난이도: lv3
// 태그: greedy
//
// 시간복잡도: O(E log E) (간선 정렬)
// 공간복잡도: O(V) (부모 배열)
package pg42861

import "sort"

func Solution(n int, costs [][]int) int {

	sort.Slice(costs, func(i, j int) bool { return costs[i][2] < costs[j][2] })
	parent := make([]int, n)
	for i := range parent {
		parent[i] = i
	}
	cost := 0
	for i := range costs {
		x, y, edge := costs[i][0], costs[i][1], costs[i][2]
		rootX := findParent(parent, x)
		rootY := findParent(parent, y)
		if rootX != rootY {
			unionParent(parent, rootX, rootY)
			cost += edge
		}
	}
	return cost
}

func findParent(parent []int, x int) int {
	if parent[x] != x {
		parent[x] = findParent(parent, parent[x])
	}
	return parent[x]
}

func unionParent(parent []int, x, y int) {
	if x < y {
		parent[x] = y
	} else {
		parent[y] = x
	}
}

File: java/src/main/java/problems/pg42861/Solution.java
package problems.pg42861;

import java.util.Arrays;
import java.util.Comparator;

/**
 * [PG] 42861 - 섬 연결하기 https://programmers.co.kr/learn/courses/30/lessons/42861 난이도: lv3 태그: greedy
 *
 * <p>시간복잡도: O(E Log E) 공간복잡도: O(N) - 부모 배열
 */
public class Solution {

  public int solution(int n, int[][] costs) {

    Arrays.sort(costs, Comparator.comparingInt(a -> a[2]));

    int[] parents = new int[n];
    for (int i = 0; i < n; i++) {
      parents[i] = i;
    }
    int answer = 0;
    int totalCount = 0;
    for (final int[] cost : costs) {
      int parentX = findParent(parents, cost[0]);
      int parentY = findParent(parents, cost[1]);
      int edge = cost[2];
      if (parentX != parentY) {
        unionParent(parents, parentX, parentY);
        answer += edge;
        totalCount += 1;
        if (totalCount == n-1) break;
      }
    }
    return answer;
  }

  private int findParent(int[] parents, int index) {
    if (parents[index] != index) {
      parents[index] = findParent(parents, parents[index]);
    }
    return parents[index];
  }

  private void unionParent(int[] parents, int parentX, int parentY) {
    if (parentX < parentY) {
      parents[parentY] = parentX;
    } else {
      parents[parentX] = parentY;
    }
  }
}
